### PR TITLE
[WIP] Add MemoryHook

### PR DESCRIPTION
**Pull request by delta2323**

### CODE REVIEW

File: chainer/function_hooks/memory.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+from chainer import cuda
+from chainer import function
+
+
+class MemoryHook(function.FunctionHook):
+    """Function hook that detects CUDA memory allocation and deallocation.
+
```
Comment (beam2d) [inline chainer/function_hooks/memory.py]: Explain what this hook actually does, and how/when to use.

File: chainer/function_hooks/memory.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        if self.cuda_available:
+            cuda.cuda.set_allocator(self.prev_allocator)
+
+    def forward_postprocess(self, function, in_data):
+        self._postprocess()
+
+    def backward_postprocess(self, function, in_data, out_grad):
+        self._postprocess()
+
+    def total_allocate_size(self):
```
Comment (beam2d) [inline chainer/function_hooks/memory.py]: Write a docstring.

File: chainer/function_hooks/memory.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        return sum(s if e == 'alloc' else 0
+                   for _, e, s in self.allocate_history)
+
+    def total_deallocate_size(self):
```
Comment (beam2d) [inline chainer/function_hooks/memory.py]: Write a docstring.

**Comment by beam2d** [comment]
Add it to the reference.

**Comment by delta2323** [comment]
I made this PR WIP because design is not satisfactory.

File: cupy/cuda/memory.pyx [hunk 4]
```diff
@@ -436,3 +449,30 @@ cdef class MemoryPool(object):
...
+cdef class ProfiledMemory(MemoryBase):
+
+    def __init__(self, memory, listener):
+        super(ProfiledMemory, self).__init__(
+            memory.size, memory.device, memory.ptr)
+
+        self._memory = memory
+        self._listener = listener
+
+    def __del__(self):
```
Comment (jheymann85) [inline cupy/cuda/memory.pyx]: `__del__` does not get called but `__dealloc__` works (at least for me)

**Comment by delta2323** [comment]
I close this PR as this is too old.

**Pull request closed without merging**